TypeORM Entity 생성 및 적용

seoko97 opened the pull request:
* Closes #13

## ✨ **구현 기능 명세**
- `User`, `Plan`, `Tag`, `Category` Entity 생성
- `User - Plan`, `User - Tag`, `User - Category` 관계 설정
- `Plan - Tag`, `Plan - Category` 관계 설정

## 🎁 **주목할 점**
- Swagger 적용을 위한 `ApiProperty` decorator 적용
- TypeORM 관계설정을 위한 `OneToMany`, `ManyToOne`, `ManyToMany` decorator 적용

seoko97 committed:
chore. update .gitignore
- add .vscode

seoko97 committed:
feat. plan entity update로 인한 mock data 수정
- test/plan/mock.ts 파일에서 생성한 MOCK_PLAN 데이터 적용

seoko97 committed:
feat. AppModule Entity 추가
- Category, Tag Entity 추가

seoko97 committed:
feat. plan entity 수정
- title, description, color, type, startTime, endTime, user, category, tags column 추가
- user, category, tags column relation 추가

seoko97 committed:
feat. Category entity 추가
- name, color, user, plans column 추가
- User entity에 대한 1:N 관계 설정
- Plan entity에 대한 1:N 관계 설정

seoko97 committed:
feat. Tag entity 생성
- name, user, plans column 생성
- User entity에 대한 1:N 관계 설정
- Plan entity에 대한 N:M 관계 설정

seoko97 committed:
feat. User entity 생성
- username, email, profileImage, categories, tags column 생성
- Category entity에 대한 1:N 관계 설정
- Tag entity에 대한 1:N 관계 설정

seoko97 committed:
feat. camel case -> snake case 변경 (#13)(#14)
- typeorm-naming-strategies 라이브러리를 통해 column명 변경

seoko97 committed:
Fix. default value 적용 안되던 오류 수정 (#13)(#14)
- Buffer를 통한 hex 형식변환으로 올바른 데이터 저장

jintak0401 commented on src/entity/category.entity.ts:
```diff
@@ -0,0 +1,37 @@
+  @Column({
+    type: 'binary',
+    length: 3,
+    default: Buffer.from('0x52d681'),
```
binary는 이렇게 하는 거였군요

pyo-sh commented on src/app.module.ts:
```diff
@@ -2,13 +2,16 @@ import { Module } from '@nestjs/common';
 import { Plan } from '@/entity/plan.entity';
 import { User } from '@/entity/user.entity';
 
 import { PlanModule } from './api/plan/plan.module';
+import { Category } from './entity/category.entity';
+import { Tag } from './entity/tag.entity';
```
절대 경로 마렵습니다... 선생님 😅

pyo-sh commented on test/api/plan/plan.service.spec.ts:
```diff
@@ -1,17 +1,12 @@
-    updatedAt: new Date('2023-03-07 15:38:06.785155'),
-    user: {} as User,
-  };
+  const mockedPlan = Object.assign(MOCK_PLAN);
```
`Object.assign`을 사용한 이유는 뭔가요???

seoko97 replied:
스프레드 연산자와 고민했는데 확실히 기존 객체를 복사한다는걸 명시해주고싶어 사용했습니다.
생각해보니 첫번째 인자에 빈객체를 통해 불변성울 지켜줘야겠네요 😭

jintak0401 commented on src/entity/plan.entity.ts:
```diff
@@ -1,27 +1,103 @@
-    onDelete: 'CASCADE',
+  @Column({ type: 'text', nullable: true })
+  @IsString()
+  @IsEmpty()
```
`@IsEmpty()` 가 아니라 `@IsOptional()`로 바꾸고 `@IsString()` 위쪽에 있어야 하는거 아닌가요?

jintak0401 commented on src/entity/plan.entity.ts:
```diff
@@ -1,27 +1,103 @@
+  @ApiProperty()
+  @Column({ type: 'boolean', default: true })
+  @IsBoolean()
+  @IsNotEmpty()
```
`@IsNotEmpty()`가 `@IsBoolean()` 위쪽에 있어야 하는거 아닌가요?

seoko97 replied:
실행은 정상적으로 실행되지만 선언적으로 생각하면 그게 맞겠네요.. 수정하게습니다.ᐟ

jintak0401 commented on src/entity/plan.entity.ts:
```diff
@@ -1,27 +1,103 @@
+  @ApiProperty()
+  @Column({ type: 'datetime', nullable: true })
+  @IsDate()
+  @IsEmpty()
```
nullable인 것들은 `@IsEmpty()`가 아니라 `@IsOptional()`로 해서 다른 class validator들보다 위쪽에 있어야 할 것 같은데 어떻게 생각하시나요?

pyo-sh replied:
`class validator decorator order`라는 키워드로 구글에 검색해 나온 [stackoverflow](https://stackoverflow.com/questions/63254004/can-i-specify-the-order-to-execute-validation-decorators-with-class-validator)를 참조했습니다.

[Typescript - decorator-composition](https://www.typescriptlang.org/docs/handbook/decorators.html#decorator-composition)에 따르면 데코레이터로 적용하는 함수들은 `Top-to-Bottom` 적용,
적용된 함수들의 실행은 `Bottom-to-Top` 적용이라고 합니다.

추가적으로 [class-validator 라이브러리](https://www.npmjs.com/package/class-validator#passing-options)를 참조해서 실행 테스트를 진행해봤습니다
```typescript
import { IsNotEmpty, IsString, validate } from "class-validator";

class MyValidator {
  @IsString()
  @IsNotEmpty()
  hi: string;
}

const a = new MyValidator();
a.hi = undefined;

validate(a).then((errors) => {
  if (errors.length > 0) {
    console.log("validation failed. errors: ", errors);
  } else {
    console.log("validation succeed");
  }
});
```

* 실행 결과
    ![image](https://user-images.githubusercontent.com/55688122/225862708-3df91ec6-ab5f-478b-b005-4f229052d50c.png)

`class-validator` 라이브러리에서는 바로 Error를 내는 것이 아니라 모든 함수 실행 후 `constraints` 안에 있는 내용을 기반으로 Error를 표현하고 있는 것을 확인할 수 있습니다! (기능적으로 문제 X)

따라서 Validate하는 과정은 Decorator 적용 순서와는 상관이 없어보입니다!
하지만 선언적인면에서는 해당 방법이 맞는 것 같네요

jintak0401 commented on src/entity/plan.entity.ts:
```diff
@@ -1,27 +1,103 @@
+  @ManyToMany(() => Tag, (tag) => tag.plans, { cascade: true })
+  @JoinTable({
+    joinColumn: { name: 'plan_id' },
+    inverseJoinColumn: { name: 'tag_id' },
```
`inverseJoinColumn`이 무엇인지 설명해주세요

seoko97 replied:
`@JoinTable()`을 통해 참조 테이블을 생성하면서 내부 column을 생성합니다. 이때 각 테이블 마다 연결될 참조 column에 관해 설정해야합니다.

- `joinColumn`: 현재 테이블의 column `(default: id)` 연결 정보 정의
- `inverseJoinColumn`:  현재 테이블과 참조관계가 설정된 테이블의 column 연결 정보 정의

jintak0401 commented on src/entity/tag.entity.ts:
```diff
@@ -0,0 +1,27 @@
+  @Column({ type: 'varchar', length: 20 })
+  @ApiProperty()
+  @IsString()
+  @IsNotEmpty()
```
`@IsNotEmpty()`가 `@IsString()` 보다 위쪽에 있어야 할 것 같습니다.

jintak0401 commented on src/entity/tag.entity.ts:
```diff
@@ -0,0 +1,27 @@
+
+@Entity('tag_tb')
+export class Tag extends DefaultEntity {
+  @Column({ type: 'varchar', length: 20 })
```
어떤 것들은 `@Column()` 이 위쪽에 있고 어떤 것들은 `@ApiProperty()`가 위쪽에 있습니다. 통일시켜주는게 어떨까요?

seoko97 committed:
Feat. Object.assign 첫번째 인자 빈객체 추가 (#14)
- 빈 객체를 넣어주면 두번째 인자의 값들이 복사되어 들어가는 것을 이용하여 기존의 객체를 수정하지 않고 새로운 객체를 만들어서 리턴하도록 수정

seoko97 committed:
feat: Entity Decorator 순서 변경 (#14)
- 선언적 프로그래밍을 위해 IsNotEmpty, IsEmpty의 순서를 최상위로 변경

seoko97 committed:
Feat. IsEmpty -> IsOptional 변경 (#14)

pyo-sh reviewed (approved):
정말 고생하셨습니다,,, 😭

seoko97 committed:
feat: ApiProperty 최상위 순서로 변경 (#14)

jintak0401 reviewed (approved):
고생하셨습니다!
